[red-knot] Add internal documentation on kinds of types (#13878)
astral-sh/ruff: crates/red_knot_python_semantic/docs/kind_of_types.md

A sealed type is a type for which it is known
that there is only a finite and pre-determined set of inhabitants at runtime.
It follows from this that for every sealed type,
there is also only a finite and pre-determined set of *subtypes* of that type.

A sealed type in Python has the property
that its proper subtypes[^2] are all disjunct from one another,
and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.

All singleton types are sealed types; however, not all sealed types are singleton types.

Examples of sealed types (other than the singleton types listed above) are:

- `bool`:

    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
    - The only proper subtypes of `bool` are `Literal[True]` and `Literal[False]`.
        `Literal[True]` and `Literal[False]` are disjunct types, and their union is
        exactly equal to `bool`.

- Enums: consider the following enum class:

    ```py
    from enum import Enum, auto

    class Foo(Enum):
        X = auto()
        Y = auto()
    ```

    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
        `Foo.X` and `Foo.Y`

>     - The only proper subtypes of `Foo` are `Literal[Foo.X]` and `Literal[Foo.Y]`.

Reviewer: … and `Never`.

        `Literal[Foo.X]` and `Literal[Foo.Y]` are disjunct, and the union `Literal[Foo.X, Foo.Y]`
        is exactly equal to the type `Foo`.

Because a singleton type is equivalent to the union of all of its proper subtypes,
for any given singleton type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
`X & ~A == B | C`. To give a Python example:

```py
from enum import Enum, auto

class X(Enum):
    A = auto()
    B = auto()
    C = auto()

def f(var: X):
    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)

    if var is X.A:
        ...  # var is narrowed to `Literal[X.A]`
    else:
        ...  # var is narrowed to `Literal[X.B, X.C]`
```

## Single-value types

A single-value type is a type for which it is known
that all inhabitants of the type are equivalent with respect to their underlying value.

> All singleton types are single-value types, but not all single-value types are singleton types.

Reviewer: One thing that could be clarified here: Is `Never` a single-value type or not? The way I implemented `is_single_valued()`, `Never` is excluded. So that definition would have to include something like "A single-value type is a non-empty type for which …"

Examples of single-value types that are not singleton types
are `Literal["foo"]`, `Literal[b"foo"]`, and `Literal[123456]`.
All inhabitants of the `Literal["foo"]` type are entirely fungible and equal
(they are all equal to the string `"foo"`).
However, they are not necessarily the *same* object in terms of identity;
multiple instances of the `"foo"` string can exist at runtime at different memory addresses.[^1]
This means that it is not safe to narrow a single-value type by identity unless it is also known

> that the type is a singleton type.

Reviewer: This is only true for narrowing using `is not`. For narrowing using `!=`, single-valuedness is the relevant property. See https://github.com/astral-sh/ruff/pull/13749

In order for a type to be considered a single-value type,
there must exist a reflexive, symmetric, and transitive equivalence relation
between all inhabitants of the type. It follows that for a single-value type in Python,
all inhabitants must be instances of the same runtime class,
and the class in question must have `__eq__` and `__ne__` methods
such that the equality relation between instances of the class
is reflexive, symmetric, and transitive.

Single-value types in unions can be safely narrowed using inequality (but not using equality):

```py
from typing import Literal

def f(x: str | Literal[1]):
    if x != 1:
        ...  # x can be narrowed to `str` if not equal to `1`
    else:
        ...  # type of x is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)

    if x == 1:
        ...  # type of x is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)
    else:
        ...  # x can be narrowed to `str` if not equal to `1`
```

## Closed types

> A closed type is a type for which it is known that the type has no subtypes except for the type itself;

Reviewer: Sorry for being pedantic: "except for the type itself and the empty type".

Author: Already fixed this in my latest push, I think :-)

Author: Also this is exactly the kind of PR where pedantry is most important haha

- A "literal class type": a type representing a single known class (and none of its possible subclasses).
- A "literal function type": a type representing a single known function
    (and excludes all other functions, even if they have the same signature).
- A "literal module type": a type representing a single known module.

Since it is known that all inhabitants of a given singleton type share the same memory address,
singleton types in unions can be safely narrowed by identity,
using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):

```py
def f(x: str | None):
    if x is None:
        ...  # x can be narrowed to `str`
    else:
        ...  # x can be narrowed to `None`
```

All singleton types are also sealed types, closed types and single-value types
(see below for definitions).

## Sealed types

A sealed type is a type for which it is known
that there is only a finite and pre-determined set of inhabitants at runtime.
It follows from this that for every sealed type,
there is also only a finite and pre-determined set of *subtypes* of that type.

A sealed type in Python has the property

> that its proper subtypes[^2] are all disjunct from one another,

Author: Great point, thanks. I think the best solution is just to remove this paragraph.

and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.

All singleton types are sealed types; however, not all sealed types are singleton types.

Examples of sealed types (other than the singleton types listed above) are:

- `bool`:

    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
        `Literal[True]`, `Literal[False]` and `Never` are all disjunct types,

>         and their union is exactly equal to `bool`.

Reviewer 2: This is true for `bool`, because it only has two inhabitants, so we can't construct any non-disjoint union types of its inhabitants.

- Enums: consider the following enum class:

    ```py
    from enum import Enum, auto

    class Foo(Enum):
        X = auto()
        Y = auto()
        Z = auto()
    ```

    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:

>         `Foo.X` and `Foo.Y`

Reviewer 2: Here you mention only `X` and `Y`, but below you also mention `Z`

    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`, and `Never`.
        `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]` and `Never` are all disjunct,

>         and the union `Literal[Foo.X, Foo.Y, Foo.Z] | Never` is exactly equal to the type `Foo`.

Reviewer 2: This is not true, because it fails to consider unions of inhabitants, which are also proper subtypes of `Foo`. But I think if you just replace "proper subtype" with "inhabitant", leave out Never, and don't mention disjointness (since individual objects are by definition "disjoint" from each other), you can reach the same conclusion below (regarding how a negated intersection with the overall type is equivalent to a union of literal types for its inhabitants), which is the important part.

Because a sealed type is equivalent to the union of all of its proper subtypes,
for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
`X & ~A == B | C`. To give a Python example:

```py
from enum import Enum, auto

class X(Enum):
    A = auto()
    B = auto()
    C = auto()

def f(var: X):
    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)

    if var is X.A:
        ...  # var is narrowed to `Literal[X.A]`
    else:
        ...  # var is narrowed to `Literal[X.B, X.C]`
```

## Single-value types

A single-value type is a non-empty type for which it is known

> that all inhabitants of the type are equivalent with respect to their runtime value.

Reviewer 2: It might be considered implied by "equivalence", but I think we should be explicit from the beginning that we are talking here about equivalence with respect to Python equality as defined by `__eq__` and `__ne__` implementations. "Equivalent with respect to their runtime value" is less clear what exactly we are talking about.

Author: The introductory paragraph at the top of the document states:

> This document provides definitions for various kinds of types, with examples of Python types that satisfy these definitions.

I'm _trying_ to present generalised definitions in the first 1-2 sentences of each section, that would apply outside of the specific context of Python, then in later sentences expand on the initial definition and link it back to the concept of Python types. So I would _prefer_ not to mention `__eq__` or `__ne__` in the first sentence here, as this is very specific to the way equivalence between arbitrary objects is generally implemented in Python.

Reviewer 2: I didn't realize you were aiming to initially present a definition of each type abstracted from the Python context. I'm not sure how concretely valuable that is, since this is documentation for red-knot, which is a Python type-checker. And I think it's a bit challenging for "single-valued type" specifically, because what we mean by the term in red-knot _is_ specific to the relation defined by `__eq__` and `__ne__`. It's certainly possible, though.

Reviewer 2: I think the core issue here is that the phrase "equivalent with respect to their runtime value" doesn't have any clear meaning in an abstract type system sense. So if you want to present an abstract definition, you'll have to start by saying "Assuming some equivalence relation is defined on type inhabitants..." and go from there.

(You can probably also skip discussion of reflexivity, transitivity, and symmetry, and relegate that to a footnoted link to e.g. https://en.wikipedia.org/wiki/Equivalence_relation, if you are using the term "equivalence relation", since an equivalence relation is defined to require all of those properties.)

Author: > I didn't realize you were aiming to initially present a definition of each type abstracted from the Python context. I'm not sure how concretely valuable that is, since this is documentation for red-knot, which is a Python type-checker.

It's not documentation for _users_ of red-knot though; it's documentation for red-knot developers to help them (us) think about type-system concepts. I personally find it helpful to consider these concepts in a more abstract way initially, before then linking the concepts back to the specific context of Python :-)

All singleton types are single-value types, but not all single-value types are singleton types.

Examples of single-value types that are not singleton types
are `Literal["foo"]`, `Literal[b"foo"]`, `Literal[123456]`, and `tuple[Literal[True], Literal[False]]`.

> All runtime inhabitants of the `Literal["foo"]` type are entirely fungible and equal

Reviewer 2: I would also suggest avoiding a description like "entirely fungible", because it is not well-defined. Whether the inhabitants are fungible or not is a matter of perspective and what you're doing with them! If you are doing `is` comparisons, they are not fungible. It's better if we state things precisely in terms of some equivalence relation: rather than "entirely fungible and equal", the type inhabitants are "all in the same equivalence class according to some equivalence relation" (the Python equality relation, in our specific case.)

Author: Sad -- there are so few situations in which you get to use the word "fungible"; I really thought this was one of them

Author: (While I concede that you're pedantically correct, I _do_ struggle to see what you might be doing with numbers such that one `50000000000000000` object would not be fungible with another `50000000000000000` object 😄)

Reviewer 2: Only things you shouldn't be doing with them, for sure! The intent of the language is absolutely that they should be fungible.

(they are all equal to the string `"foo"`).
However, they are not necessarily the *same* object in terms of identity;
multiple instances of the `"foo"` string can exist at runtime at different memory addresses.[^1]
This means that it is not safe to narrow a single-value type by identity unless it is also known
that the type is a singleton type.

In order for a type to be considered a single-value type,
there must exist a reflexive, symmetric, and transitive equivalence relation

> between all inhabitants of the type. It follows that for a single-value type in Python,

Reviewer 2: The definition of a single-value type is specific to the Python equality relation, so it seems odd to say "there must exist a ... relation" here.

I would say instead that a single-value type must have a reflexive, symmetric, and transitive definition of Python equality.

But really I kind of think this entire paragraph might not be necessary; I think the literal-type examples you give above are probably sufficient to clarify what a real single-value type looks like. In practice we are never going to try to define any single-value type on a non-builtin class, or any class that might challenge the requirements.

Reviewer 2: > `Literal[E.A]`, for an enum `E` with a member `A`, is a single-value type.

Fair! And if we observe a custom `__eq__` implementation on an `Enum` subclass, we probably have to stop treating literal types of those enum members as single-valued types, because we no longer can claim to understand the equality relation for instances of that enum type.

all inhabitants must be instances of the same runtime class,
and the class in question must have `__eq__` and `__ne__` methods
such that the equality relation between instances of the class
is reflexive, symmetric, and transitive.

Single-value types in unions can be safely narrowed using inequality (but not using equality):

```py
from typing import Literal

def f(x: str | Literal[1]):
    if x != 1:
        ...  # `x` can be narrowed to `str` if not equal to `1`
    else:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)

    if x == 1:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)
    else:
        ...  # `x` can be narrowed to `str` if not equal to `1`
```

## Closed types

A closed type is a type for which it is known that the type has no subtypes
except for the type itself and `Never`; it is "closed for extension".
For a given closed type `X`, the only subtypes of `X` are `X` and `Never`;

> the only *proper* subtype of `X` is `Never`.

Reviewer 2: I think this wording has a similar problem with arbitrary intersection and union types as the above wording on sealed types.

I think we have to present closed types in a way that's more tied to "instance" types -- as in, it's a type defined as instances of a class, which cannot be subclassed, therefore all inhabitants of the type must be instances of the final class.

Author: > I think we have to present closed types in a way that's more tied to "instance" types -- as in, it's a type defined as instances of a class, which cannot be subclassed, therefore all inhabitants of the type must be instances of the final class.

I'm not sure we want the concept of "closed" types to be limited to _just_ `@final` nominal types, though. I think we at least want `TypedDict`s with `closed=True` to qualify (assuming [PEP 728](https://peps.python.org/pep-0728/) is accepted)?

Closed types are often associated with the `@final` decorator.
Decorating a class with `@final` indicates to the type checker
that the class should never be subclassed; the type checker will emit a diagnostic
if it sees it being subclassed, and is free to consider the class as closed for extension.
As such, the following `@final` class `Spam` creates a corresponding closed type `Spam`:

```py
from typing import final

@final
class Spam: ...
```

However, runtime subclassability does not correspond exactly to whether a type is closed or not.
For example, the `bool` *class* is (correctly) decorated with `@final` in typeshed's stubs,
indicating that it cannot be subclassed at runtime. However, the `bool` *type*
has two proper subtypes other than `Never`: `Literal[True]` and `Literal[False]`.
As such, although all its subtypes are closed, `bool` itself cannot be considered a closed type.
Similarly, attempting to subclass the `Eggs` class in the following example
will lead to an exception at runtime, and type checkers should treat all
enum classes as implicitly `@final`. Nonetheless, the `Eggs` *type* cannot be considered closed,
as it has three proper subtypes other than `Never`:

> `Literal[Eggs.A]`, `Literal[Eggs.B]` and `Literal[Eggs.C]`.

Reviewer 2: I don't think this paragraph is right. I don't think "has no proper subtypes other than itself and Never" is a workable or useful definition of what it means to be a closed type, and I think both `bool` and `Eggs` should be considered closed types.

Author: What _is_ the term you would use for a type where it is known that no proper subtypes exist other than `Never`?

Reviewer 2: I think that would have to be a singleton type. Any type with more than one inhabitant in principle has proper subtypes that are not `Never`.

Reviewer 2: None that we have (or ever would) define or attempt to name, but sure, in principle you can have the singleton type consisting only of the integer object with value `50000000000000000` that exists at a certain memory address, which is a proper subtype of `Literal[50000000000000000]` that is not `Never`.

Author: grrr, fine!

Author: In some ways I find this a useful clarification: it's good to remember that the set of Python types recognised by red-knot is a subset of the set of possible types that exist in Python. From this perspective, there really exists an infinite number of possible singleton types in Python.

In other ways, I find this not so useful, because I still feel like we're missing a term to describe types which _in red-knot's model_ can have no possible non-`Never` proper subtypes _that would be understood as such by red-knot_. You've successfully convinced me that this is a concept that cannot be expressed in terms of an abstract framework divorced from the concept of Python types and our internal model of Python types. But I still think it's a useful concept nonetheless and something we should have a term for.

Reviewer 2: I think I'm not understanding _why_ this is a useful or important concept.

Author: It is what I was _trying_ to describe... but perhaps you're right, maybe it just isn't actually a useful thing to describe. Not sure...

Anyway, I'm reworking the section titled closed types to more accurately describe the actual type-theory concept described by the term "closed types" :-) will probably finish off that rewrite tomorrow morning and push those changes.

Reviewer 2: Ok, looking forward to it!

Reviewer: > > `Literal[50000000000000000]` is not a singleton type (multiple inhabitants exist at runtime), but has no non-`Never` proper subtypes
>
> None that we have (or ever would) define or attempt to name, but sure, in principle you can have the singleton type consisting only of the integer object with value `50000000000000000` that exists at a certain memory address, which is a proper subtype of `Literal[50000000000000000]` that is not `Never`.

This is purely a discussion about nomenclature, so I'm not sure how useful it is to go deeper, but I'm not sure I completely agree with this *"every possible set of runtime values corresponds to *some* type, even if we can't name/construct it"* viewpoint. I think the following definition is more useful: A type is a syntactic form that we assign to terms of the programming language. In this definition, it's still true that every type represents a set of possible runtime values. But it's not true that every possible set of runtime values corresponds to a type. *Which* particular sets of runtime values can be represented as a type depends on the type system and its expressiveness. In C's type system, you can not assign a type to the set `{true}`. In Python, you can; In Python, you can not assign a type to objects that are used exactly once. In languages with linear types, you can.

In Python's type system, there is no way to construct a type that represents all objects at memory location 0xf7…, and I don't think it's useful to think of types in terms of sets of runtimes values that could *potentially* be represented by a more expressive type system. I would not claim that C has intersection types, just because I can think of a set of runtime values that would be represented by `int8_t & uint8_t = {0, 1, …, 127}`. And I would not claim that Python has linear types, just because I can think of the set of runtime objects that are used exactly once.

Author: > This is purely a discussion about nomenclature

Hey, the whole PR is about nomenclature 😆

One complication here is that red-knot's model of Python's types _is_ more expressive than the model outlined by the typing spec. (In fact, you could argue that all Python type checkers necessarily have a more expressive model than the one outlined in the typing spec -- but I think red-knot goes further than most here.) One obvious example is our concept of function-literal and class-literal types -- you _could_ argue that the concept of class-literal types in particular is strongly implied by much of the rest of the typing spec, but nowhere is it explicitly specified. Similarly, we treat `IntersectionType`s as first-class types and infer them in many places -- but an `Intersection` type is still unspecified in the typing spec. (Mypy, for example, [does narrowing in many places that clearly _implies_ the existence of an intersection type](https://github.com/python/mypy/blob/60d1b3776229d3e3c332413ef49ad89be469a5e4/mypy/checker.py#L5416-L5440), but it has no first-class notion of intersection types in its model.)

So I could put forward at least three possible definitions of "type" here:
- A maximally broad definition: "Any theoretical set of runtime objects constitutes a 'type'"
- A maximally narrow definition: "Only a type explicitly specified by the typing spec can be described as a type"
- A maximally useful definition: "Any set of runtime objects constitutes a type if it would plausibly be useful to represent it as a type in red-knot in order to accurately and precisely model Python's runtime semantics"

I think the maximally useful definition is probably the most, well... useful... but it's also the least precise! It's hard to know which sets of runtime objects we might plausibly find useful to represent as types _at some point_.

Reviewer 2: Once a type system has full set-theoretic types (arbitrary unions, intersections, and negations), it becomes quite hard to draw a clear line between which possible sets of objects are a type and which are not; many quite complex types can be constructed "out of thin air". Because of this, and because of what Alex says (which sets of objects we choose to represent as a type internally is not set in stone by either the language or the typing spec, but is essentially up to our own judgment of what is useful for representing Python semantics, and may change over time), I think the clearest approach in our terminology is to avoid defining types based on an assumption that no subtype of the type could ever exist (despite the type having multiple distinguishable inhabitants.)

(But perhaps more important to the original intent of this thread is that I also don't find that definition _otherwise_ useful or correct: I do think that `bool` and `Eggs` are closed types.)

Author: > I do think that `bool` and `Eggs` are closed types

Yeah, you're right that I was trying to use the word "closed" to describe a different concept to what it means everywhere else in type theory; that was obviously wrong and silly. No disagreements on _that_ point.

        A = auto()
        B = auto()
        C = auto()
    ```

- A "literal class type": a type representing a single known class (and none of its possible subclasses).

- A "literal function type": a type that represents a single known function
    (and excludes all other functions, even if they have the same signature).

- A "literal module type": a type representing a single known module.

Since it is known that all inhabitants of a given singleton type share the same memory address,
singleton types in unions can be safely narrowed by identity,
using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):

```py
def f(x: str | None):
    if x is None:
        ...  # x can be narrowed to `str`
    else:
        ...  # x can be narrowed to `None`
```

All singleton types are also sealed types, and closed types; nearly all are single-value types.
(See below for definitions of these other concepts.)

The number of singleton types in Python is theoretically infinite
(for any given theoretical runtime object `x`, you could theorise a singleton type `Y` for which
the sole inhabitant of `Y` would be the runtime object `x`). However, only a small fraction of

> the set of possible singleton types are recognised as being singleton types by red-knot's model.

Reviewer 2: This phrasing feels slightly off to me. It's not about whether red-knot "recognizes them as singleton types" -- it's not as if red-knot recognizes them as some other kind of type instead! It's just that they are types for which red-knot has no internal representation (nor is there any way to spell them in a type expression.)

there is also only a finite and pre-determined set of *subtypes* of that type.

All singleton types are sealed types; however, not all sealed types are singleton types.

Examples of sealed types (other than the singleton types listed above) are:

- `bool`:

    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
    - The only proper subtypes[^2] of `bool` are `Literal[True]`, `Literal[False]`, and `Never`;
        their union is exactly equal to `bool`.

- Enums: consider the following enum class:

    ```py
    from enum import Enum, auto

    class Foo(Enum):
        X = auto()
        Y = auto()
        Z = auto()
    ```

    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
        `Foo.X`, `Foo.Y` and `Foo.Z`.

    - The only proper subtypes of `Foo` are:

        - `Literal[Foo.X]`
        - `Literal[Foo.Y]`
        - `Literal[Foo.Z]`
        - `Literal[Foo.X, Foo.Y]`
        - `Literal[Foo.X, Foo.Z]`
        - `Literal[Foo.Y, Foo.Z]`
        - `Never`

        The union of the proper subtypes of `Foo` (`Literal[Foo.X, Foo.Y, Foo.Z] | Never`)
        is exactly equal to the type `Foo`.

> Because a sealed type is equivalent to the union of all of its proper subtypes,

Reviewer 2: It's not clear to me what the importance is here of the description "equivalent to the union of all of its proper subtypes" -- that describes every type! Not every type has a finite and enumerable set of proper subtypes, but _every_ type is equivalent to the union of all its proper subtypes.

> for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,

Reviewer 2: ```suggestion
for any given sealed type `X` where the only inhabitants of `X` are `A`, `B` and `C`,
```

Author: No, I think that would make the sentence inaccurate. Your phrasing is true for the specific case of enum members, which I use as an example in the sentences following this, but is not true in the abstract, and this sentence describes an abstract relation. Here I theorise a hypothetical sealed type where there are exactly three proper subtypes and describe how such a sealed type might work.

I think it's possibly confusing that I'm using the same letters in my abstract description of the relation as I then use in my enum example immediately below, since the enum example does not in fact correspond exactly to the abstract relation I'm describing here

`X & ~A == B | C`. To give a Python example:

```py
from enum import Enum, auto

class X(Enum):
    A = auto()
    B = auto()
    C = auto()

def f(var: X):
    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)

    if var is X.A:
        ...  # var is narrowed to `Literal[X.A]`
    else:
        ...  # var is narrowed to `Literal[X.B, X.C]`
```

## Single-value types

For a given type `T`, `T` can be said to be a single-value type if there exists an
[equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation)
between all inhabitants of the type and the equivalence relation is satisfied between
all inhabitants of the type. For a Python type to be categorised as a single-value type,
all inhabitants of the type must be instances of the same runtime class,
and the class in question must have `__eq__` and `__ne__` methods
such that the equality relation between instances of the class

> is reflexive, symmetric, and transitive.

Reviewer 2: This seems to leave out the key third requirement: that the inhabitants of the type are all equal to each other according to the `__eq__` and `__ne__` implementations of that class. It was stated above in the abstract form, but for clarity it seems worth saying it again in the specific form.

1. Definitions that would be useful for developers working on red-knot.

## Types

A static Python type represents a set of possible values at runtime.
The number of ways in which possible Python objects could be categorised into sets and subsets is infinite;
thus, there is an infinite number of possible types in Python.

The [typing spec](https://typing.readthedocs.io/en/latest/spec/) specifies various kinds of types
that all Python type checkers should support,
The desirable type checker behavior has been standardised to some degree for these specified types,
but the spec is by no means exhaustive.
There are more kinds of types in Python than are included in the spec,
and many kinds of types that are left unspecified must be understood by a type checker
in order for the type checker to accurately and precisely model Python's runtime semantics.

Some types in Python have a known size.
Most, however, have an infinite number of possible inhabitants and subtypes.
For any type `T`, the union of all of the proper subtypes[^2] of `T` is exactly equal to `T`.

## Singleton types

A singleton type is a type for which it is known that there is

> (and can only ever be) exactly one known inhabitant of the type at runtime:

Reviewer 2: nit: the use of the qualifier "known" in this sentence invites uncertainty about whether you are saying "there is definitely exactly one inhabitant, and it is known" or "there is exactly one known inhabitant, there may be other unknown ones."

You could be more explicit that it is the former, but I'm not sure it's worth it, vs just removing the word "known" -- I think it's obvious that if we can be sure there's exactly one inhabitant of type, that means we know what it is :)
```suggestion
(and can only ever be) exactly one inhabitant of the type at runtime:
```

a set of runtime values with size exactly 1.
For any singleton type in Python, the type's sole inhabitant will always exist
at the same memory address for the entire duration of a Python program.

Examples of singleton types in red-knot's model of Python include:

- `types.EllipsisType` (the sole inhabitant is `builtins.Ellipsis`;
    the constructor always returns the same object):

>     ```pycon

Reviewer 2: ooh TIL that this is a usable code block language, neat!

- `Literal[True]`: the sole inhabitant of the type is the constant `True`.

- `Literal[False]`: the sole inhabitant of the type is the constant `False`.

- `Literal[E.A]`, `Literal[E.B]` or `Literal[E.C]` for the following enum `E`
    (the sole inhabitant of `Literal[E.A]` is the enum member `E.A`):

    ```py
    from enum import Enum, auto

    class E(Enum):
        A = auto()
        B = auto()
        C = auto()
    ```

- A "literal class type": a type representing a single known class (and none of its possible subclasses).

- A "literal function type": a type that represents a single known function
    (and excludes all other functions, even if they have the same signature).

- A "literal module type": a type representing a single known module.

Since it is known that all inhabitants of a given singleton type share the same memory address,
singleton types in unions can be safely narrowed by identity,
using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):

```py
def f(x: str | None):
    if x is None:
        ...  # x can be narrowed to `str`
    else:
        ...  # x can be narrowed to `None`
```

> All Python singleton types are also sealed types and final types; nearly all are single-value types.

Reviewer 2: There's still a reference to "final type" here, which may no longer make sense given later changes.

For a given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
`X & ~A == B | C`.

Due to the enumerability of a sealed type's inhabitants,
a Python sealed type can be narrowed using identity:

```py
from enum import Enum, auto

class Colors(Enum):
    RED = auto()
    BLUE = auto()
    YELLOW = auto()

def f(var: X):
    ...  # the type of `var` here is `X` (which is equivalent to `Literal[Colors.RED, Colors.BLUE, Colors.YELLOW]`)

    if var is Colors.RED:
        ...  # var is narrowed to `Literal[Colors.RED]`
    else:
        ...  # var is narrowed to `Literal[Colors.BLUE, Colors.YELLOW]`
```

## Single-value types

For a given type `T`, `T` can be said to be a single-value type if there exists an
[equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation)
between all inhabitants of the type and the equivalence relation is satisfied between
all inhabitants of the type.

For a Python type `T` to be categorised as a single-value type:

- All inhabitants of `T` must be instances of the same runtime class `U`
- `U` must have `__eq__` and `__ne__` methods such that the equality relation between instances of `U`

>     is reflexive, symmetric, and transitive.

Reviewer 2: Not important, but may help readers tie together some mentioned concepts.

```suggestion
    is reflexive, symmetric, and transitive (satisfying the definition of an equivalence relation.)
```

- For any two inhabitants of `T` named `x` and `y`, `x == y`.

Nearly all singleton types are single-value types, but the reverse is not true.
Many single-value types exist that are not singleton types: examples include
`Literal[123456]`, `Literal[b"foo"]`, `tuple[Literal[True], Literal[False]]`, and `Literal["foo"]`.
All runtime inhabitants of the `Literal["foo"]` type are equal to the string `"foo"`.

> However, they are not necessarily the *same* object in terms of identity;

Reviewer 2: nit: I think "in terms of identity" is redundant here, especially when you immediately go on to specifically discuss memory addresses
```suggestion
However, they are not necessarily the *same* object;
```

multiple `str` instances equal to `"foo"` can exist at runtime at different memory addresses.[^1]
This means that it is not safe to narrow a single-value type by identity unless it is also known

> that the type is a singleton type.

Reviewer 2: We said above that we can also narrow sealed types by identity, not just singleton types:
```suggestion
that the type is a sealed type.
```

Single-value types in unions can be safely narrowed using inequality (but not using equality):

```py
from typing import Literal

def f(x: str | Literal[1]):
    if x != 1:
        ...  # `x` can be narrowed to `str` if not equal to `1`
    else:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)

    if x == 1:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)
    else:
        ...  # `x` can be narrowed to `str` if not equal to `1`
```

An example of a singleton type that is not a single-value type
would be `Literal[Ham.A]` for the following `Ham` enum:
because the `Ham` class overrides `__eq__`, we can no longer say for sure
that the `Ham` type can be safely narrowed using inequality as in the above example

> (but we *can* still narrow the `Ham` type using identity, as with all singleton types):

Reviewer 2: `Ham` is not a singleton type, but `Literal[Ham.A]` is

```suggestion
(but we *can* still narrow the `Literal[Ham.A]` type using identity, as with all singleton types):
```

```py
from enum import Enum, auto

class Ham(Enum):
    A = auto()
    B = auto()
    C = auto()

    def __eq__(self, other):
        return False
```

## Final classes

A "final class" in Python is a class that can be statically inferred as being unsubclassable.
Most final classes are decorated with [`@typing.final`](https://docs.python.org/3/library/typing.html#typing.final),
a special decorator that indicates to the type checker that the tool should assume
the class cannot be subclassed, and that the tool should emit an error
if it detects an attempt to do so. Not all final classes are decorated with `@final`, however.
The most common example of this is enum classes. Any enum class that has at least one member is considered
implicitly final, as attempting to subclass such a class will fail at runtime.

> For any two classes `X` and `Y`, if `X` is `@final` and `X` is not a subclass of `Y`,

Reviewer 2: nit: this applies to all final classes, whether or not they are made final by the use of `@final` (and saying "is `@final`" reads oddly to me, even as a way to say "is decorated with `typing.final`")
```suggestion
For any two classes `X` and `Y`, if `X` is final and `X` is not a subclass of `Y`,
```

the intersection of the types `X & Y` is equivalent to `Never`, the empty set.
This therefore means that for a type defined as being "all instances of the final class `X`",

> there are fewer ways of subtyping such a type than for most types in Python.

Reviewer 2: I still feel like this section buries the lede in a major way.

From red-knot's perspective, by far the most important thing about final class types is that we know exactly what methods and attributes they define, with what signatures (including especially dunder methods), and we know there can't be infinite possible subclasses overriding those methods and attributes. But we don't even mention that here, beyond a very vague reference in this phrase to "fewer ways of subtypign"; instead we go on to discuss (at relative length, with an example) the edge case of how some final class types can still have proper subtypes, although it remains unclear in the text why the existence or non-existence of proper subtypes should matter to red-knot or the reader.

> However, types associated with final classes can still have non-`Never` proper subtypes.

Reviewer 2: I think "types associated with final classes" is a vague and verbose way to word this; "final class types" I think is a clearer unambiguous reference to the "a type defined as ..." in the previous sentence.
```suggestion
However, final class types can still have non-`Never` proper subtypes.
```

Author: I guess I was going out of my way to avoid using the term "final types" here, because I was getting less and less convinced that it was actually a useful term that didn't just invite confusion 😄 But you're probably right that my roundabout language here doesn't really add any clarity.

Enums again provide a good example here: `Literal[Animals.LION]` is a proper subtype
of the `Animals` type here, despite the fact that the `Animals` class cannot be subclassed:

```py
from enum import Enum, auto

> class Animals(Enum):

Reviewer 2: minor nit, not related to the core point: I think using plurals is generally not great naming practice for either enums or classes generally, suggest `Animal` instead

    LION = auto()
    LEOPARD = auto()
    CHEETAH = auto()
```

Many singleton and sealed types are associated with final classes: the classes
`types.EllipsisType`, `types.NotImplementedType`, `types.NoneType`, `bool`,
and all enum classes are final classes. However, there are also many singleton
and sealed types (some representable, but some merely theoretical)
that cannot be defined as "all instances of a specific final class `X`".
In the above enum, `Literal[Animals.LION]` is a singleton type and a sealed type,

> but would not fit this definition.

Reviewer 2: I would probably eliminate this entire paragraph. It's confusing, and really not clear what important point it makes.

## Types

A static Python type represents a set of possible values at runtime.
The number of ways in which possible Python objects could be categorised into sets and subsets is infinite;
thus, there is an infinite number of possible types in Python.[^3]

The [typing spec](https://typing.readthedocs.io/en/latest/spec/) specifies various kinds of types
that all Python type checkers should support,
The desirable type checker behavior has been standardised to some degree for these specified types,
but the spec is by no means exhaustive.
There are more kinds of types in Python than are included in the spec,
and many kinds of types that are left unspecified must be understood by a type checker
in order for the type checker to accurately and precisely model Python's runtime semantics.

A runtime object in Python can be said to "inhabit" a static type if the object satisfies
the conditions by which the type is defined. For example, the `str` type can be defined
as "all runtime objects which are instances of the class `builtins.str`".
According to this definition, the runtime strings `"foo"` and `"bar"`

> both inhabit the `str` type.

Author: Yes. Though I'm not sure this information exists in the typing spec exactly as written. But I was also worried about overlap. I'll take another look.

> both inhabit the `str` type.

Reviewer 2: https://typing.readthedocs.io/en/latest/spec/concepts.html#fully-static-types says

> If an object v is a member of the set of objects denoted by a fully static type T, we can say that v is a “member of” the type T, or v “inhabits” T.

Single-value types in unions can be safely narrowed using inequality (but not using equality):

```py
from typing import Literal

def f(x: str | Literal[1]):
    if x != 1:
        ...  # `x` can be narrowed to `str` if not equal to `1`
    else:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)

    if x == 1:
        ...  # type of `x` is still `str | Literal[1]`
             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
             # to compare equal with `1`)
    else:
        ...  # `x` can be narrowed to `str` if not equal to `1`
```

An example of a singleton type that is not a single-value type
would be `Literal[Ham.A]` for the following `Ham` enum:
because the `Ham` class overrides `__eq__`, we can no longer say for sure
that the `Ham` type can be safely narrowed using inequality as in the above example
(but we *can* still narrow the `Ham` type using identity, as with all singleton types):

```py
from enum import Enum, auto

class Ham(Enum):
    A = auto()
    B = auto()
    C = auto()

    def __eq__(self, other):

>         return True

Reviewer 2: It may be confusing to readers that technically `Literal[Ham.A]` still is a single-value type, given this definition of `Ham`, because all `Ham` instances compare equal to everything! And this is technically still a valid equivalence relation, in which all inhabitants of `Literal[Ham.A]` do compare equal.

I think in  _practice_ though, we will not want to give red-knot that level of intelligence about the specific implementation of `__eq__`; we will rather want to just say that if `__eq__` is defined, we don't understand the equality relation, and can no longer conclude that `Literal[Ham.A]` is a single-valued type.

It may be useful to clarify this distinction, or it may be fine to just change this `return True` to `return False` and avoid the issue altogether here.

```

## Final types

A final type in Python is a [nominal type](https://docs.python.org/3/library/typing.html#nominal-vs-structural-subtyping)
defined as the instances of a certain runtime class that it is known cannot be subclassed.
Most final types in Python are associated with class definitions that are decorated with
[`@typing.final`](https://docs.python.org/3/library/typing.html#typing.final).
This is a special decorator which indicates to the type checker that the tool
should assume the class cannot be subclassed, and that the tool should emit an error
if it detects an attempt to do so.

There are fewer ways in which final types can be subtyped when compared to most other types.
For any two nominal types `X` and `Y`, if `X` is `@final` and `X` is not a subtype of `Y`,
the intersection `X & Y` is equivalent to `Never`, the empty set.

> However, a final type in Python still has subtypes, just no *nominal* subtypes.

Reviewer 2: Nit: are you talking about proper subtypes here? Every final type has one nominal subtype: itself!

Author: Argh, yes! And `Never` is a subtype of everything, again, of course

Author: > And `Never` is a subtype of everything, again, of course

...Except `Never` is not a _nominal_ subtype.

For the following two types `X` and `Truthy`, the intersection `X & Truthy` is non-empty,
since `Truthy` is a structural type:

```py
from typing import final, Protocol, Literal

class Truthy(Protocol):
    """Any object `x` inhabits this type if `bool(x)` always returns `True`"""

    def __bool__(self) -> Literal[True]: ...

@final
class X[T: bool]:
    def __init__(self, bool_val: T):
        self.bool_val = bool_val

    def __bool__(self) -> T:
        return self.bool_val

> ```

Reviewer 2: I think this section is a bit debatable.

If we are talking about a conceptual `Truthy` type consisting of "all objects whose `__bool__` actually returns `True`", that might work for this example, but it is probably not a good example to include since it's not clear if we even want to define such a type, given that individual objects can easily change their `__bool__` behavior over time.

If we are talking about the Protocol type shown here, then given a final type X, `X & Truthy` will always be either `X` or `Never`, it will never be a proper subtype of `X`, because `X.__bool__` has a known signature, and that signature is either a subtype of `(self) -> Literal[True]` or it is not.

A better example of proper subtypes of final types might be a final generic type with a covariant type parameter? In that case `X[B] <: X[A]` if `B <: A`, even though `X` is final.

At a higher level, though, it remains unclear to me why it is important to discuss whether a final type has any proper subtypes here?

Author: I find it pretty useful to remember that runtime subclassing and static subtyping are linked but not the same thing. Just because something is `@final` does not mean it cannot be subtyped, only that it cannot be subclassed; all types except singleton types can be subtyped.

Not all final types are decorated with `@final`. The most common example of this
is enum classes. Any enum class that has at least one member is considered

> implicitly final, as attempting to subclass such a class will fail at runtime:

Reviewer 2: Is there a missing code example following this? Or should the colon be a full stop instead?

Author: Oh oops, I added a code example and then deleted it again because it didn't really add anything. I'll get rid of the colon.
